modules/afsocket/afinet-dest.c

```diff
@@ -492,65 +492,88 @@ afinet_dd_construct_ipv6_packet(AFInetDestDriver *self, LogMessage *msg, GString
-    {
-      gboolean success = FALSE;
+  if (!self->lnet_buffer)
+    self->lnet_buffer = g_string_sized_new(self->spoof_source_maxmsglen);
 
-      g_assert(self->super.transport_mapper->sock_type == SOCK_DGRAM);
+  log_writer_format_log(self->super.writer, msg, self->lnet_buffer);
 
-      g_static_mutex_lock(&self->lnet_lock);
+  if (self->lnet_buffer->len > self->spoof_source_maxmsglen)
+    g_string_truncate(self->lnet_buffer, self->spoof_source_maxmsglen);
 
-      if (!self->lnet_buffer)
-        self->lnet_buffer = g_string_sized_new(self->spoof_source_maxmsglen);
+  gboolean success = afinet_dd_construct_ip_packet(self, msg, self->lnet_buffer);
 
-      log_writer_format_log(self->super.writer, msg, self->lnet_buffer);
+  if (!success)
+    goto finish;
 
-      if (self->lnet_buffer->len > self->spoof_source_maxmsglen)
-        g_string_truncate(self->lnet_buffer, self->spoof_source_maxmsglen);
+  success = libnet_write(self->lnet_ctx) >= 0;
+
+  if (!success)
+    {
+      msg_error("Error sending raw frame", evt_tag_str("error", libnet_geterror(self->lnet_ctx)));
+      goto finish;
+    }
+
+  /* we have finished processing msg */
+  log_msg_ack(msg, path_options, AT_PROCESSED);
+  log_msg_unref(msg);
+
+finish:
+  g_static_mutex_unlock(&self->lnet_lock);
+  return success;
+}
+
+static inline gboolean
+_is_message_spoofable(LogMessage *msg)
+{
+  return msg->saddr && (msg->saddr->sa.sa_family == AF_INET || msg->saddr->sa.sa_family == AF_INET6);
+}
+
+static inline gboolean
+_is_spoof_source_enabled(AFInetDestDriver *self)
+{
+  return self->spoof_source && self->lnet_ctx;
+}
 
-      switch (self->super.dest_addr->sa.sa_family)
-        {
-        case AF_INET:
-          success = afinet_dd_construct_ipv4_packet(self, msg, self->lnet_buffer);
-          break;
-#if SYSLOG_NG_ENABLE_IPV6
-        case AF_INET6:
-          success = afinet_dd_construct_ipv6_packet(self, msg, self->lnet_buffer);
-          break;
 #endif
-        default:
-          g_assert_not_reached();
-        }
-      if (success)
-        {
-          if (libnet_write(self->lnet_ctx) >= 0)
-            {
-              /* we have finished processing msg */
-              log_msg_ack(msg, path_options, AT_PROCESSED);
-              log_msg_unref(msg);
 
-              g_static_mutex_unlock(&self->lnet_lock);
-              return;
-            }
-          else
-            {
-              msg_error("Error sending raw frame",
-                        evt_tag_str("error", libnet_geterror(self->lnet_ctx)));
-            }
-        }
-      g_static_mutex_unlock(&self->lnet_lock);
+static void
+afinet_dd_queue(LogPipe *s, LogMessage *msg, const LogPathOptions *path_options)
+{
+#if SYSLOG_NG_ENABLE_SPOOF_SOURCE
+  AFInetDestDriver *self = (AFInetDestDriver *) s;
+
+  /* NOTE: this code should probably become a LogTransport instance so that
+   * spoofed packets are also going through the LogWriter queue */
+
```

Comment:
Just a reply to the NOTE in comment: the LogTransport interface does not carry source address information, thus it is not possible to do that. We could do a similar thing at the LogProto layer, which still gets the LogMessage instance.